crates/red_knot_python_semantic/src/types/property_tests.rs

```diff
@@ -0,0 +1,238 @@
+        Ty::KnownClassInstance(KnownClass::Bool),
+        Ty::KnownClassInstance(KnownClass::List),
+        Ty::KnownClassInstance(KnownClass::Tuple),
+        Ty::KnownClassInstance(KnownClass::FunctionType),
+        Ty::KnownClassInstance(KnownClass::SpecialForm),
+        Ty::KnownClassInstance(KnownClass::TypeVar),
+        Ty::KnownClassInstance(KnownClass::TypeAliasType),
+        Ty::KnownClassInstance(KnownClass::NoDefaultType),
+        Ty::TypingLiteral,
+        Ty::BuiltinClassLiteral("str"),
+        Ty::BuiltinClassLiteral("int"),
+        Ty::BuiltinClassLiteral("bool"),
+        Ty::BuiltinClassLiteral("object"),
+    ])
+    .unwrap()
+    .clone()
+}
+
+/// Constructs an arbitrary type.
+///
+/// The `size` parameter controls the depth of the type tree. For example,
+/// a simple type like `int` has a size of 0, `Union[int, str]` has a size
+/// of 1, `tuple[int, Union[str, bytes]]` has a size of 2, etc.
+fn arbitrary_type(g: &mut Gen, size: u32) -> Ty {
+    if size == 0 {
+        arbitrary_core_type(g)
+    } else {
+        match u32::arbitrary(g) % 4 {
+            0 => arbitrary_core_type(g),
+            1 => Ty::Union(
+                (0..*g.choose(&[2, 3]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+            ),
+            2 => Ty::Tuple(
+                (0..*g.choose(&[0, 1, 2]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+            ),
+            3 => Ty::Intersection {
+                pos: (0..*g.choose(&[0, 1, 2]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+                neg: (0..*g.choose(&[0, 1, 2]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl Arbitrary for Ty {
+    fn arbitrary(g: &mut Gen) -> Ty {
+        const MAX_SIZE: u32 = 2;
+        arbitrary_type(g, MAX_SIZE)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        // This is incredibly naive. We can do much better here by
+        // trying various subsets of the elements in unions, tuples,
+        // and intersections. For now, we only try to shrink by
+        // reducing unions/tuples/intersections to a single element.
+        match self.clone() {
+            Ty::Union(types) => Box::new(types.into_iter()),
+            Ty::Tuple(types) => Box::new(types.into_iter()),
+            Ty::Intersection { pos, neg } => Box::new(pos.into_iter().chain(neg)),
+            _ => Box::new(std::iter::empty()),
+        }
+    }
+}
+
+static CACHED_DB: OnceLock<Arc<Mutex<TestDb>>> = OnceLock::new();
+
+fn get_cached_db() -> MutexGuard<'static, TestDb> {
+    let db = CACHED_DB.get_or_init(|| Arc::new(Mutex::new(setup_db())));
+    db.lock().unwrap()
+}
+
+macro_rules! type_property_test {
```

Comment:
I find the macro fairly hard to read. What I understand is that their primary benefit is to reduce the repetition of creating the cached db and dereferencing it. 

* How much faster is reusing the db over creating a new database?
* Have you considered adding a snapshot method to the database instead of locking it? Databases are thread safe and can be shared (for as long as you don't update the db with a `&mut Db` reference

```
        pub fn snapshot(&self) -> Self {
            Self {
                storage: self.storage.clone(),
                files: self.files.snapshot(),
                system: self.system.clone(),
                vendored: self.vendored.clone(),
                events: Arc::clone(&self.events),
            }
        }
```

A test then becomes 

```
    #[test]
    #[quickcheck_macros::quickcheck]
    fn equivalent_to_is_reflexive2(t: crate::types::tests::Ty) -> bool {
        let db = get_cached_db();
        let t = t.into_type(&db);
        t.is_equivalent_to(&db, t)
    }
```

Which I don't find too bad

Reply:
> How much faster is reusing the db over creating a new database?

Three orders of magnitude. It's essential. I found some bugs only after I added this optimization.

> * Have you considered adding a snapshot method to the database instead of locking it

I have not. Would the benefit be to get rid of the `let db = &*db;`, or do you think that would be faster?

> I find the macro fairly hard to read

Because macros are hard to read, or because this particular macro is hard to read? :smile:

I don't disagree, but it's also not like we're hiding lots of complexity in there. It's really just a boilerplate-reduction macro. I personally care much more about the readability of the individual tests. The macro is ugly, yes. But I won't have to read the macro code again when writing new tests.

> A test then becomes [...] Which I don't find too bad

Well, you picked the easiest example there ;-). With just one type and no logical implication. Did you see the macro-vs-non-macro comparison in the PR description? Or this example, which has three types:

```rs
#[quickcheck]
fn subtype_of_is_transitive(s: Ty, t: Ty, u: Ty) -> bool {
    let db = get_cached_db();
    let db = &*db;
    let s = s.into_type(db);
    let t = t.into_type(db);
    let u = u.into_type(db);

    !(s.is_subtype_of(db, t) && t.is_subtype_of(db, u)) || s.is_subtype_of(db, u)
}
```

```rs
type_property_test!(
    subtype_of_is_transitive,
    db,
    (s, t, u),
    s.is_subtype_of(db, t) && t.is_subtype_of(db, u) => s.is_subtype_of(db, u)
);
```

> their primary benefit is to reduce the repetition of creating the cached db and dereferencing it

It also removes the necessity for lots of `let t = t.into_type(db);` lines, and allows us to write logical implications using `premise => conclusion` instead of `!(premise) || (conclusion)`.

Let me know what you think. I'm going to be okay if we remove it again :smile:.

Reply:
> I have not. Would the benefit be to get rid of the let db = &*db;, or do you think that would be faster?

It would allow you to remove the mutex guard and the deref (you still need to use `&db` when calling any function)


> Because macros are hard to read, or because this particular macro is hard to read? 😄

It's somewhat both. I had to expand the macro to understand what the different parts map to. Like what's up with the `(s, t, u)`? which part is even the assertion? Some extra documentation might help. 


> Did you see the macro-vs-non-macro comparison in the PR description? Or this example, which has three types:

I admit, I did not! I see how it simplifies the code a bit, but it doesn't significantly reduce the number of lines. 

Could we implement `into_type` for tuples so that you can keep writing `(s, t, u) = (s, t, u).into_types(db)`?

> It also removes the necessity for lots of let t = t.into_type(db); lines, and allows us to write logical implications using premise => conclusion instead of !(premise) || (conclusion).

I do like that. Maybe we can limit the macro part to just that? 

I'm not opposed to keeping the macros. I just had to expand them, and I still spent about 5 minutes trying to figure out what was going on. But I'm also not sure if it's worth it if you spend more time on it.

Reply:
> Like what's up with the `(s, t, u)`? which part is even the assertion?

These are valid concerns. I now pushed another commit where I change the syntax to the following:

```rs
type_property_test!(
    subtype_of_implies_assignable_to, db,
    forall types s, t. s.is_subtype_of(db, t) => s.is_assignable_to(db, t)
);
```

The intention is that it reads like a mathematical property `∀ S, T. …`.

I also simplified the macro definition (only two clauses instead of four), clarified the variable names and added some documentation.

I kind of like how this looks like, and would merge it for now. I'm happy to revisit it though if I get more feedback on it (from you or others).



> It would allow you to remove the mutex guard and the deref (you still need to use `&db` when calling any function)

I'll note it down as a task for me (low priority), thanks.

Reply:
Thanks. I like this much more!